zcash_client_backend/src/proto.rs

```diff
@@ -183,3 +204,223 @@ impl service::TreeState {
         read_commitment_tree::<Node, _, NOTE_COMMITMENT_TREE_DEPTH>(&sapling_tree_bytes[..])
     }
 }
+
+pub const PROPOSAL_SER_V1: u32 = 1;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ProposalError<DbError> {
+    Zip321(Zip321Error),
+    TxIdInvalid(Vec<u8>),
+    InputRetrieval(DbError),
+    InputNotFound(TxId, PoolType, u32),
+    BalanceInvalid,
+    MemoInvalid(memo::Error),
+    VersionInvalid(u32),
+    ZeroMinConf,
+    FeeRuleNotSpecified,
+}
+
+impl<E> From<Zip321Error> for ProposalError<E> {
+    fn from(value: Zip321Error) -> Self {
+        Self::Zip321(value)
+    }
+}
+
+impl proposal::ProposedInput {
+    pub fn parse_txid(&self) -> Result<TxId, Vec<u8>> {
+        Ok(TxId::from_bytes(self.txid.clone().try_into()?))
+    }
```

Comment:
Could use at least token documentation. Also instead of cloning, use the `TryFrom<&[T]>` impl that returns [TryFromSliceError](https://doc.rust-lang.org/std/array/struct.TryFromSliceError.html) instead of the vec:
```suggestion
    pub fn parse_txid(&self) -> Result<TxId, TryFromSliceError> {
        Ok(TxId::from_bytes(self.txid[..].try_into()?))
    }
```